### ISSUE TITLE
DST should be array of bytes.

### ISSUE THREAD

**Issue by armfazh** [paulmillr/noble-curves#16]
Regarding the parameters of `hash_to_field`:

```js
DST: string, 
p: bigint, 
m: number, 
k: number, 
expand: 'xmd' | 'xof',
hash: H,
```

DST must be an array of bytes.
There should be a way to validate that H is conforming with the expander, so it avoid to call `expander = "xof" and hash = "sha-256"` for example.

https://github.com/paulmillr/noble-curves/blob/0163b635324dbf503d4ff2670e23ec2fb55e1620/src/abstract/hash-to-curve.ts#L121-L140

**Comment by paulmillr** [comment]
> There should be a way to validate that H is conforming with the expander, so it avoid to call expander = "xof" and hash = "sha-256" for example.

Don't understand, can you clarify this?

**Comment by armfazh** [comment]
xmd only applies to Merkle-Damagard type of functions (SHA family for example)
xof are only extendable output functions (SHAKE for example).
it should be not possible to call `expander='xof'` with `H = 'sha-256'`

**Comment by paulmillr** [comment]
We don't use strings in function description, we use interfaces. Users are free to define their own methods conforming to interfaces.

This is done for tree shaking: if you use sha256, no sha3 code is imported for you.

So instead of passing "sha256", you need to pass sha256 (function) which conforms to CHash interface.

Not sure how we can filter xofs with that.

**Comment by paulmillr** [comment]
Regarding dst: string I will change it to string | Uint8Array.

Keeping strings a second option is important for UX.

**Comment by armfazh** [comment]
> We don't use strings in function description, we use interfaces. Users are free to define their own methods conforming to interfaces.
> 
> This is done for tree shaking: if you use sha256, no sha3 code is imported for you.
> 
> So instead of passing "sha256", you need to pass sha256 (function) which conforms to CHash interface.

I understand that you are not using strings, but my point is not that.
My point is that whatever you pass on the `H` object, must be the implementation of either Merkle-Damagard functions only when expand='xmd`; or the implementation of a XOF when passing expand='xof'. Other combinations are not allowed.

> Not sure how we can filter xofs with that.

Not sure if the type system can help here.

> Regarding dst: string I will change it to string | Uint8Array.
> Keeping strings a second option is important for UX.

I recommend you to use the same types for `msg` and `dst`, so both array of bytes. if the user needs to pass a string they can always use TextEncoder to convert the parameter.

**Comment by paulmillr** [comment]
> either Merkle-Damagard functions only when expand='xmd`

Non-shake Sha3 (which is not Merkle Damagard) is also allowed, right?

**Comment by paulmillr** [comment]
> I recommend you to use the same types for msg and dst, so both array of bytes. if the user needs to pass a string they can always use TextEncoder to convert the parameter.

Have you seen any case where dst is bytes? Glancing over spec it seems ALL dsts there are human readable strings (even if they are converted to bytes then).

**Comment by armfazh** [comment]
> Non-shake Sha3 (which is not Merkle Damagard) is also allowed, right?

Yes.

> Have you seen any case where dst is bytes? Glancing over spec it seems ALL dsts there are human readable strings (even if they are converted to bytes then).

From the spec:

> The example tags given above are assumed to be ASCII-encoded byte strings without null termination, which is the RECOMMENDED format. Other encodings can be used, but in all cases the encoding as a sequence of bytes MUST be specified unambiguously.

Thinking twice, strings in Javascript are Unicode, not sure if this will be an issue. Maybe not.

**Comment by paulmillr** [comment]
Fixed in 6bc4b35. For xof stuff we'll try to fix it later once we get types in place.
